doc/man/man5/sched_conf.5
'\" t
.\"___INFO__MARK_BEGIN__
.\"
.\" Copyright: 2004 by Sun Microsystems, Inc.
.\"
.\"___INFO__MARK_END__
.\" $RCSfile: sched_conf.5,v $     Last Update: $Date: 2009/07/08 14:42:40 $     Revision: $Revision: 1.39 $
.\"
.\"
.\" Some handy macro definitions [from Tom Christensen's man(1) manual page].
.\"
.de SB		\" small and bold
.if !"\\$1"" \\s-2\\fB\&\\$1\\s0\\fR\\$2 \\$3 \\$4 \\$5
..
.\"
.de T		\" switch to typewriter font
.ft CW		\" probably want CW if you don't have TA font
..
.\"
.de TY		\" put $1 in typewriter font
.if t .T
.if n ``\c
\\$1\c
.if t .ft P
.if n \&''\c
\\$2
..
.\"
.de M		\" man page reference
\\fI\\$1\\fR\\|(\\$2)\\$3
..
.TH SCHED_CONF 5 "$Date: 2009/07/08 14:42:40 $" "xxRELxx" "xxQS_NAMExx File Formats"
.\"
.SH NAME
sched_conf \- xxQS_NAMExx default scheduler configuration file
.\"
.\"
.SH DESCRIPTION
.I sched_conf
defines the configuration file format for xxQS_NAMExx's  scheduler. 
In order to modify the configuration, 
use the graphical user's interface
.M qmon 1
or the
.I -msconf
option of the 
.M qconf 1
command. A default configuration is provided together with the 
xxQS_NAMExx distribution package.
.PP
Note, xxQS_NAMExx allows backslashes (\\) be used to escape newline
(\\newline) characters. The backslash and the newline are replaced with a
space (" ") character before any interpretation.
.\"
.\"
.SH FORMAT
The following parameters are recognized by the xxQS_NAMExx scheduler if
present in \fIsched_conf\fP:
.SS "\fBalgorithm\fP"
.B Note:
Deprecated, may be removed in future release.
.br
Allows for the selection of alternative scheduling algorithms.
.PP
Currently
.B default
is the only allowed setting.
.\"
.SS "\fBload_formula\fP"
A simple algebraic expression used to derive a single weighted
load value from all or part of the load parameters reported by
.M xxqs_name_sxx_execd 8
for each host and from all or part of the consumable resources (see
.M complex 5 )
being maintained for each host.
The load formula expression syntax is that of
a summation weighted load values, that is:
.sp 1
.nf
.RS
{w1|load_val1[*w1]}[{+|-}{w2|load_val2[*w2]}[{+|-}...]]
.RE
.fi
.sp 1
\fBNote\fP, no blanks are allowed in the load formula.
.br
The load values and consumable resources (load_val1, ...)
are specified by the name defined in the complex (see
.M complex 5 ).
.br
.B Note:
Administrator defined load values (see the
.B load_sensor
parameter in
.M xxqs_name_sxx_conf 5
for details)
and consumable resources available for all hosts (see
.M complex 5 )
may be used as well as xxQS_NAMExx default load parameters.
.br
The weighting factors (w1, ...) are positive integers. After the expression
is evaluated for each host the results are assigned to the hosts and
are used to sort the hosts corresponding to the weighted load. The sorted
host list is used to sort queues subsequently.
.br
The default load formula is "np_load_avg".
.SS "\fBjob_load_adjustments\fP"
The load, which is imposed by the xxQS_NAMExx jobs 
running on a system varies in time, and often, e.g. for the CPU load, 
requires some amount of time to be reported in the appropriate 
quantity by the operating system. Consequently, if a job was started 
very recently, the reported load may not provide a sufficient 
representation of the load which is already imposed on that host by 
the job. The reported load will adapt to the real load over time, but 
the period of time, in which the reported load is too low, may 
already lead to an oversubscription of that host. xxQS_NAMExx allows 
the administrator to specify \fBjob_load_adjustments\fP which are used 
in the xxQS_NAMExx scheduler to compensate for this problem.
.br
The \fBjob_load_adjustments\fP are specified as a comma separated list
of arbitrary load parameters or consumable resources and (separated by an
equal sign) an
associated load correction value. Whenever a job is dispatched to a
host by the scheduler,
the load parameter and consumable value set of that host
is increased by the values
provided in the \fBjob_load_adjustments\fP list. These correction
values are decayed linearly over time until after 
\fBload_adjustment_decay_time\fP from the start the corrections
reach the value 0.
If the \fBjob_load_adjustments\fP
list is assigned the special denominator NONE, no load corrections are
performed.
.br
The adjusted load and consumable values are used to compute the
combined and weighted
load of the hosts with the \fBload_formula\fP (see above) and to compare
the load and consumable values against the load threshold lists
defined in the queue configurations (see
.M queue_conf 5 ).
If the \fBload_formula\fP consists simply of the default CPU load average 
parameter \fInp_load_avg\fP, and if the jobs are very compute intensive, one might
want to set the \fBjob_load_adjustments\fP list to \fInp_load_avg=1.00\fP,
which means that every new job dispatched to a host will require
100 % CPU time, and thus the machine's load is instantly increased by 1.00.
.SS "\fBload_adjustment_decay_time\fP"
The load corrections in the "\fBjob_load_adjustments\fP" list above
are decayed linearly over time from the point of the job start, where
the corresponding load or consumable parameter is
raised by the full correction value,
until after a time period of "\fBload_adjustment_decay_time\fP", where the
correction becomes 0. Proper values for "\fBload_adjustment_decay_time\fP"
greatly depend upon the load or consumable parameters used and the
specific operating
system(s). Therefore, they can only be determined on-site and experimentally.
For the default \fInp_load_avg\fP load parameter a
"\fBload_adjustment_decay_time\fP" of 7 minutes has proven to yield reasonable
results.
.SS "\fBmaxujobs\fP"
The maximum number of jobs any user may have running in a xxQS_NAMExx
cluster at the same time. If set to 0 (default) the users may run an arbitrary
number of jobs. 
.SS "\fBschedule_interval\fP"
At the time the scheduler thread initially registers at the event master thread in 
.M xxqs_name_sxx_qmaster 8 process
\fBschedule_interval\fP is used to set the time interval in which
the event master thread 
sends scheduling event updates to the scheduler thread.
A scheduling event is a status change that has occurred within
.M xxqs_name_sxx_qmaster 8
which may trigger or affect scheduler decisions (e.g. a job has
finished and thus the allocated resources are available again).
.br
In the xxQS_NAMExx default scheduler the arrival of
a scheduling event report triggers a scheduler run. The scheduler
waits for event reports otherwise.
.br
\fBSchedule_interval\fP is a time value (see
.M queue_conf 5
for a definition of the syntax of time values).
.SS "\fBqueue_sort_method\fP"
This parameter determines in which order several criteria are taken into
account to product a sorted queue list. Currently, two settings are valid:
\fBseqno\fP and \fBload\fP. However in both cases, xxQS_NAMExx attempts to
maximize the number of soft requests (see
.M qsub 1 
\fB\-s\fP option) being fulfilled by the queues for a particular as the
primary criterion.
.br
Then, if the \fBqueue_sort_method\fP parameter is set to \fBseqno\fP,
xxQS_NAMExx will use the
.B seq_no
parameter as configured in the current queue configurations (see
.M queue_conf 5 )
as the next criterion to sort the queue list. The 
.B load_formula
(see above) has only a meaning if two queues have equal
sequence numbers.
If 
.B queue_sort_method
is set to \fBload\fP the load according the 
.B load_formula
is the criterion after maximizing a job's soft requests and the sequence
number is only used if two hosts have the same load.
The sequence number sorting is most 
useful if you want to define a fixed order in which queues are to be filled
(e.g.   the cheapest resource first).
.PP
The default for this parameter is \fBload\fP.
.\"
.SS "\fBhalftime\fP"
When executing under a share based policy, the scheduler "ages" (i.e. decreases)
usage to implement a sliding window for achieving the share entitlements
as defined by the share tree. The \fBhalftime\fP defines
the time interval in which accumulated usage will have been decayed
to half its original value. Valid values are specified in hours or according to 
the time format as specified in
.M queue_conf 5 .
.br
If the value is set to 0, the usage is not decayed.
.\"
.SS "\fBusage_weight_list\fP"
xxQS_NAMExx accounts for the consumption of the resources CPU-time, memory and IO
to determine the usage which is imposed on a system by a job. A single
usage value is computed from these three input parameters by multiplying
the individual values by weights and adding them up. The weights are
defined in the \fBusage_weight_list\fP. The format of the list is
.sp 1
.nf
.RS
cpu=wcpu,mem=wmem,io=wio
.RE
.fi
.sp 1
where wcpu, wmem and wio are the configurable weights. The weights are real
number. The sum of all tree weights should be 1.
.\"
.SS "\fBcompensation_factor\fP"
Determines how fast xxQS_NAMExx should compensate for past usage below of above
the share entitlement defined in the share tree. Recommended values are
between 2 and 10, where 10 means faster compensation.
.\"
.SS "\fBweight_user\fP"
The relative importance of the user shares in the functional policy.
Values are of type real.
.\"
.SS "\fBweight_project\fP"
The relative importance of the project shares in the functional policy.
Values are of type real.
.\"
.SS "\fBweight_department\fP"
The relative importance of the department shares in the
functional policy. Values are of type real.
.\"
.SS "\fBweight_job\fP"
The relative importance of the job shares in the
functional policy. Values are of type real.
.\"
.SS "\fBweight_tickets_functional\fP"

Changelog
################################################################################
# This is the Changelog for Grid Engine source code changes.                   #
# Possible keywords: "Bugfix", "Belongs to", "Review", "Enhancem.",            #
#                    "Merged from", "Cleanup", "Issue"                         #
#                                                                              #
#                ===== PLEASE DO NOT DELETE THIS HEADER ====                   #
################################################################################
          1         2         3         4         5         6         7
01234567890123456789012345678901234567890123456789012345678901234567890123456789

EB-2009-11-14-0: Bugfix:      o JJSV related bugfixed
                                 - passing of commandline arguments was 
                                   incorrect in Java JSV script so that class 
                                   could not be found.
                                 - corrected parsing of ENV JSV command
                                 - JJSV can now accept variables without value
                                 - added code that handles special characters 
                                   contained in environment variable values
                                 - removed code that automaticaly modifys jobs
                                   if some parameters where changed although
                                   JJSV did not contain the jsv_correct command
                                 - Changed module tests accordingy
                              o Added new Java JSV script that is used in
                                jsv_script TS test
                 Review:      TBD

DT-2009-11-13-0: Bugfix:      o Fixed problems found during additional Herd
                                manual testing
                              o Added hadoop to setfileperm.sh

MP-2009-11-13-1: Bugfix:      shadowd uninstallation does not remove local 
                              host configuration
                 Bugster:     6828316
                 Issue:       2989
                 Review:      LP

MP-2009-11-13-0: Bugfix:      fixed spelling mistakes in all man pages

DG-2009-11-13-0: Enhancem.:   - Added threads in topology string in case of hyper-
                              threading on Linux architectures.
                              - Some clean ups and bug fixes
                 Review:      SH, EB

RN-2009-11-13-1: Bugfix:      qsub -sync does not return correct exit status
							  during qmaster failover
                 Bugster:     6888406
                 Issue:       3178
                 Review:      CR

JG-2009-11-13-0: Bugfix:      Fixed memory leaks in test programs and in 
                              jsv script
                 Review:      CR, EB

DT-2009-11-12-0: Bugfix:      o Fixed problems found during Herd manual testing
                              - Will rerun tests tomorrow

EB-2009-11-11-0: Bugfix:      o fixed qstat -j output that caused
                                problem in TS qstat test due to J2C binding
                                changes
                              o changed qrsh related code in shepherd
                 Review:      JG (HP)

DT-2009-11-10-2: Bugfix:      o Added SISSL headers to the remaining
                                properties files

DT-2009-11-10-1: Bugfix:      o Added SISSL headers to some properties files
                              o Added tests for Herd
                              o Fixed Herd issues found by tests
                              o Corrected error in Herd README

HP-2009-11-10-0: Bugfix:      Added SISSL headers to HTML and SDA files

DT-2009-11-10-0: Bugfix:      o Added SISSL headers to build files

DG-2009-11-10-0: Bugfix:      Fixed hardware thread reporting on Solaris.
                 Review:      CR

DT-2009-11-09-1: Bugfix:      o Added missing hadoop README

DG-2009-11-09-0: Enhancem.:   Updated showq binary in order to support core 
                              binding functionality.
                 Review:      SH

DT-2009-11-09-0: Bugfix:      o Fixed minor issue with Herd build.xml

-----------------+------------+-------------------------------------------------
MD-2009-11-09-2: CVS:		Created V62u5alpha2_TAG
-------------------------------------------------------

MD-2009-11-09-1: Bugfix:      o Fixed broken hadoop build.xml
                 Enhancem:    o added version string

MD-2009-11-09-0: Bugfix:      o Fixed broken JSV build

DT-2009-11-08-1: Bugfix:      o Fixed broken JSV build

DT-2009-11-08-0: Enhancem.:   o Added JavaDocs for Herd
                              o Addressed review comments
                              o Fixed minor issues

source/libs/spool/classic/sched_conf.c
#/*___INFO__MARK_BEGIN__*/
/*************************************************************************
 * 
 *  The Contents of this file are made available subject to the terms of
 *  the Sun Industry Standards Source License Version 1.2
 * 
 *  Sun Microsystems Inc., March, 2001
 * 
 * 
 *  Sun Industry Standards Source License Version 1.2
 *  =================================================
 *  The contents of this file are subject to the Sun Industry Standards
 *  Source License Version 1.2 (the "License"); You may not use this file
 *  except in compliance with the License. You may obtain a copy of the
 *  License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
 * 
 *  Software provided under this License is provided on an "AS IS" basis,
 *  WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
 *  WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
 *  MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
 *  See the License for the specific provisions governing your rights and
 *  obligations concerning the Software.
 * 
 *   The Initial Developer of the Original Code is: Sun Microsystems, Inc.
 * 
 *   Copyright: 2001 by Sun Microsystems, Inc.
 * 
 *   All Rights Reserved.
 * 
 ************************************************************************/
/*___INFO__MARK_END__*/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#include "sge_unistd.h"
#include "sge.h"
#include "sgermon.h"
#include "sge_gdi_request.h"
#include "sge_usageL.h"
#include "sge_schedd_conf.h"
#include "config.h"
#include "sched_conf.h"
#include "read_object.h"
#include "sge_feature.h"
#include "sge_log.h"
#include "setup_path.h"
#include "msg_common.h"
#include "spool/classic/msg_spoollib_classic.h"
#include "sge_stdio.h"
#include "sge_schedd.h"
#include "sge_spool.h"
#include "sge_string.h"
#include "sge_prog.h"
#include "sge_io.h"
#include "sge_answer.h"
#include "sge_conf.h"
#include "sge_centry.h"

static int intprt_as_load_adjustment[] = { CE_name, CE_stringval, 0 };

static int intprt_as_usage[] = { UA_name, UA_value, 0 };

static const char *delis[] = {"=", ",", "\n"};

static int str2qsm(const char *qsm_str);

static char *qsm2str(u_long32 qsm_val);

static int read_schedd_conf_work(lList **alpp, lList **clpp, int fields[], 
                                 lListElem *ep, int spool, int flag, int *tag, 
                                 int parsing_type);


/****** spool/classic/write_sched_configuration() ****************************
*  NAME
*     write_sched_configuration() -- print scheduler configuration 
*
*  SYNOPSIS
*     char* write_sched_configuration(int spool, int how, 
*                                     const char *common_dir, lListElem *ep) 
*
*  FUNCTION
*     This function is used to print the current scheduler configuration
*     in human readable form either to stdout or into a spoolfile or tmpfile. 
*
*  INPUTS
*     int spool - which fields should be written 
*        1 - write for spooling
*        2 - write only user controlled fields
*
*     int how - destination of the write operation 
*        0 - use stdout
*        1 - write into tmpfile
*        2 - write into spoolfile (use absolute path
*            common_dir/sched_configuration)
*
*     const char *common_dir - absolute path to common directory,
*                              used if how == 2
*
*     lListElem *ep - scheduler configuration (SC_Type)
*
*  RESULT
*     char* 
*        how == 0: NULL
*        how != 0: filename or NULL in case of an error
*******************************************************************************/
char *write_sched_configuration(int spool, int how, const char *common_dir, const lListElem *ep) 
{
/* JG: suppress READ_DANGLING. sge_root comes from a getenv() call.
 *     this should be handled properly in underlying function, e.g. by
 *     strdupping the value returned by getenv().
 */
   FILE *fp; 
   char fname[SGE_PATH_MAX], real_fname[SGE_PATH_MAX];
   int fret;
   dstring ds;
   char buffer[256];

   DENTER(TOP_LAYER, "write_sched_configuration");

   sge_dstring_init(&ds, buffer, sizeof(buffer));
   switch (how) {
   case 0:
      fp = stdout;
      break;
   case 1:
   case 2:
      if (how==1) {
         dstring tmp_name_error = DSTRING_INIT;
         if (sge_tmpnam(fname, &tmp_name_error) == NULL) {
            if (sge_dstring_get_string(&tmp_name_error) != NULL) {
               CRITICAL((SGE_EVENT, sge_dstring_get_string(&tmp_name_error)));
            } else {
               CRITICAL((SGE_EVENT, MSG_TMPNAM_GENERATINGTMPNAM));
            }
            sge_dstring_free(&tmp_name_error);
            DEXIT;
            return NULL;
         }
         sge_dstring_free(&tmp_name_error);
      } else if(how == 2) {
         sprintf(fname, "%s/.%s", common_dir, SCHED_CONF_FILE);
         sprintf(real_fname, "%s/%s", common_dir, SCHED_CONF_FILE);
      }

      fp = fopen(fname, "w");
      if (!fp) {
         CRITICAL((SGE_EVENT, MSG_ERRORWRITINGFILE_SS, fname, strerror(errno)));
         DEXIT;
         return NULL;
      }
      break;
   default:
      DEXIT;
      return NULL;
   }  

   if (spool && (sge_spoolmsg_write(fp, COMMENT_CHAR,
             feature_get_product_name(FS_VERSION, &ds)) < 0)) {
      goto FPRINTF_ERROR;
   }  

   /* conf values needed for both SGE & SGEEE */
   FPRINTF((fp, "algorithm                        %s\n", lGetString(ep, SC_algorithm)));
   FPRINTF((fp, "schedule_interval                %s\n", lGetString(ep, SC_schedule_interval)));
   FPRINTF((fp, "maxujobs                         " sge_u32 "\n", lGetUlong(ep, SC_maxujobs)));
   FPRINTF((fp, "queue_sort_method                %s\n", qsm2str(lGetUlong(ep, SC_queue_sort_method))));
   FPRINTF((fp, "job_load_adjustments             "));
   fret = uni_print_list(fp, NULL, 0, lGetList(ep, SC_job_load_adjustments), intprt_as_load_adjustment, delis, 0);
   if (fret < 0) {
      goto FPRINTF_ERROR;
   }
   FPRINTF((fp, "load_adjustment_decay_time       %s\n", lGetString(ep, SC_load_adjustment_decay_time)));
   FPRINTF((fp, "load_formula                     %s\n", lGetString(ep, SC_load_formula)));
   FPRINTF((fp, "schedd_job_info                  %s\n", lGetString(ep, SC_schedd_job_info)));
   FPRINTF((fp, "flush_submit_sec                 " sge_u32 "\n", lGetUlong(ep, SC_flush_submit_sec)));
   FPRINTF((fp, "flush_finish_sec                 " sge_u32 "\n", lGetUlong(ep, SC_flush_finish_sec)));
   FPRINTF((fp, "params                           %s\n", lGetString(ep, SC_params)));
   
   FPRINTF((fp, "reprioritize_interval            %s\n", lGetString(ep, SC_reprioritize_interval)));
   FPRINTF((fp, "halftime                         " sge_u32 "\n", lGetUlong(ep, SC_halftime)));
   FPRINTF((fp, "usage_weight_list                "));
   fret = uni_print_list(fp, NULL, 0, lGetList(ep, SC_usage_weight_list), intprt_as_usage, delis, 0);
   if (fret < 0) {
      goto FPRINTF_ERROR;
   }
   FPRINTF((fp, "compensation_factor              %.10g\n", lGetDouble(ep, SC_compensation_factor)));
   FPRINTF((fp, "weight_user                      %.10g\n", lGetDouble(ep, SC_weight_user)));
   FPRINTF((fp, "weight_project                   %.10g\n", lGetDouble(ep, SC_weight_project)));
   FPRINTF((fp, "weight_department                %.10g\n", lGetDouble(ep, SC_weight_department)));
   FPRINTF((fp, "weight_job                       %.10g\n", lGetDouble(ep, SC_weight_job)));
   FPRINTF((fp, "weight_tickets_functional        " sge_u32 "\n", lGetUlong(ep, SC_weight_tickets_functional)));
   FPRINTF((fp, "weight_tickets_share             " sge_u32 "\n", lGetUlong(ep, SC_weight_tickets_share)));

   FPRINTF((fp, "share_override_tickets           %s\n", lGetBool(ep, SC_share_override_tickets)?"true":"false"));
   FPRINTF((fp, "share_functional_shares          %s\n", lGetBool(ep, SC_share_functional_shares)?"true":"false"));
   FPRINTF((fp, "max_functional_jobs_to_schedule  " sge_u32 "\n", lGetUlong(ep, SC_max_functional_jobs_to_schedule)));
   FPRINTF((fp, "report_pjob_tickets              %s\n", lGetBool(ep, SC_report_pjob_tickets)?"true":"false"));
   FPRINTF((fp, "max_pending_tasks_per_job        " sge_u32 "\n", lGetUlong(ep, SC_max_pending_tasks_per_job)));
   FPRINTF((fp, "halflife_decay_list              %s\n", lGetString(ep, SC_halflife_decay_list)));
   FPRINTF((fp, "policy_hierarchy                 %s\n", lGetString(ep, SC_policy_hierarchy)));
   FPRINTF((fp, "weight_ticket                    %.10g\n", lGetDouble(ep, SC_weight_ticket)));
   FPRINTF((fp, "weight_waiting_time              %.10g\n", lGetDouble(ep, SC_weight_waiting_time)));
   FPRINTF((fp, "weight_deadline                  %.10g\n", lGetDouble(ep, SC_weight_deadline)));
   FPRINTF((fp, "weight_urgency                   %.10g\n", lGetDouble(ep, SC_weight_urgency)));
   FPRINTF((fp, "weight_priority                  %.10g\n", lGetDouble(ep, SC_weight_priority)));
   FPRINTF((fp, "max_reservation                  " sge_u32 "\n", lGetUlong(ep, SC_max_reservation)));
   FPRINTF((fp, "default_duration                 %s\n", lGetString(ep, SC_default_duration)));

   if (how != 0) {
      FCLOSE(fp);
   }

   if (how == 2 || how == 3) {
      if (rename(fname, real_fname) == -1) {
         DEXIT;
         return NULL;
      } else {
         strcpy(fname, real_fname);
      }   
   } 
   DEXIT;
 
   /* JG: TODO: ERROR: fname is returned, but is a stack variable!! */
   return how==1?sge_strdup(NULL, fname):fname;
FPRINTF_ERROR:
FCLOSE_ERROR:
   DEXIT;
   return NULL;
}

/* mapping func for SC_queue_sort_method uval -> str */
static char *qsm2str(u_long32 qsm_val) {
   switch (qsm_val) {
   case QSM_SEQNUM:
      return "seqno";
   default: 
      return "load";
   }
}

/* mapping func for SC_queue_sort_method str -> uval */
static int str2qsm(const char *qsm_str) 
{
   if (!qsm_str)
      return -1;

   if (!strcasecmp(qsm_str, "load"))
      return QSM_LOAD;
   if (!strcasecmp(qsm_str, "seqno"))
      return QSM_SEQNUM;

   return -1; /* error */
}


/****
 **** read_schedd_conf_work
 ****
 ****/
static int read_schedd_conf_work(lList **alpp, lList **clpp, int fields[], 
                                 lListElem *ep, int spool, int flag, int *tag,
                                 int parsing_type) {
   const char *str;
   lList *alp = NULL;
   u_long32 ul;
   
   DENTER(TOP_LAYER, "read_schedd_conf_work");

   /* --------- SC_halftime */
   if (!set_conf_ulong(alpp, clpp, fields, "halftime", ep, SC_halftime)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_usage_weight_list */
   if (!set_conf_deflist(alpp, clpp, fields, "usage_weight_list", ep, 
            SC_usage_weight_list, UA_Type, intprt_as_usage)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_compensation_factor */
   if (!set_conf_double(alpp, clpp, fields, "compensation_factor", ep, 
            SC_compensation_factor, 0)) {
      DEXIT;
      return -1;
   }
   
   /* --------- SC_weight_user */
   if (!set_conf_double(alpp, clpp, fields, "weight_user", ep, 
            SC_weight_user, 0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_project */
   if (!set_conf_double(alpp, clpp, fields, "weight_project", ep, 
            SC_weight_project, 0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_share_override_tickets */
   if (!set_conf_bool(alpp, clpp, fields, "share_override_tickets", ep, 
            SC_share_override_tickets)) {
      DEXIT;
      return -1;
   }
   
   /* --------- SC_share_functional_shares */
   if (!set_conf_bool(alpp, clpp, fields, "share_functional_shares", ep, 
            SC_share_functional_shares)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_report_pjob_tickets*/
   if (!set_conf_bool(alpp, clpp, fields, "report_pjob_tickets", ep, 
            SC_report_pjob_tickets)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_max_functional_jobs_to_schedule*/
   if (!set_conf_ulong(alpp, clpp, fields, "max_functional_jobs_to_schedule", ep, 
            SC_max_functional_jobs_to_schedule)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_max_pending_tasks_per_job */
   if (!set_conf_ulong(alpp, clpp, fields, "max_pending_tasks_per_job", ep, 
            SC_max_pending_tasks_per_job)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_halflife_decay_list */
   if (!set_conf_string(alpp, clpp, fields, "halflife_decay_list", ep, 
            SC_halflife_decay_list)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_department */
   if (!set_conf_double(alpp, clpp, fields, "weight_department", ep, 
            SC_weight_department,0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_job */
   if (!set_conf_double(alpp, clpp, fields, "weight_job", ep, 
            SC_weight_job, 045)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_tickets_functional */
   if (!set_conf_ulong(alpp, clpp, fields, "weight_tickets_functional", ep, 
                        SC_weight_tickets_functional)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_tickets_share */
   if (!set_conf_ulong(alpp, clpp, fields, "weight_tickets_share", ep, 
                        SC_weight_tickets_share)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_ticket */
   if (!set_conf_double(alpp, clpp, fields, "weight_ticket", ep, 
            SC_weight_ticket, 0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_waiting_time */
   if (!set_conf_double(alpp, clpp, fields, "weight_waiting_time", ep, 
            SC_weight_waiting_time, 0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_deadline */
   if (!set_conf_double(alpp, clpp, fields, "weight_deadline", ep, 
            SC_weight_deadline, 0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_urgency */
   if (!set_conf_double(alpp, clpp, fields, "weight_urgency", ep, 
            SC_weight_urgency, 0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_weight_priority */
   if (!set_conf_double(alpp, clpp, fields, "weight_priority", ep, 
            SC_weight_priority, 0)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_reprioritize_interval*/
   if (!set_conf_timestr(alpp, clpp, fields, "reprioritize_interval", ep, 
                           SC_reprioritize_interval)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_policy_hierarchy */
   if (!set_conf_string(alpp, clpp, fields, "policy_hierarchy", ep,
                        SC_policy_hierarchy)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_schedule_interval */
   if (!set_conf_timestr(alpp, clpp, fields, "schedule_interval", ep, 
                              SC_schedule_interval)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_algorithm */
   if (!set_conf_string(alpp, clpp, fields, "algorithm", ep, SC_algorithm)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_maxujobs */
   if (!set_conf_ulong(alpp, clpp, fields, "maxujobs", ep, SC_maxujobs)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_flush_submit_sec*/
   if (!set_conf_ulong(alpp, clpp, fields, "flush_submit_sec", ep, 
            SC_flush_submit_sec)) {
      DEXIT;
      return -1;
   }    
  
   if (!set_conf_string(alpp, clpp, fields, "params", ep,
            SC_params)) {
      DEXIT;
      return -1;
   }
  
   /* --------- SC_flush_finish_sec*/
   if (!set_conf_ulong(alpp, clpp, fields, "flush_finish_sec", ep, 
            SC_flush_finish_sec)) {
      DEXIT;
      return -1;
   }   

   /* --------- SC_queue_sort_method */
   str = get_conf_value(&alp, *clpp, CF_name, CF_value, "queue_sort_method");
   lFreeList(&alp);
   if (str) {
      ul = str2qsm(str);
      if (ul == (u_long32) -1) {
         SGE_ADD_MSG_ID(sprintf(SGE_EVENT, MSG_SCHEDCONF_INVALIDVALUEXFORQUEUESORTMETHOD_S, str));
         answer_list_add(alpp, SGE_EVENT, STATUS_ESYNTAX, ANSWER_QUALITY_ERROR);
         DEXIT;
         return -1; 
      }
      lSetUlong(ep, SC_queue_sort_method, ul);
      lDelElemStr(clpp, CF_name, "queue_sort_method");
      add_nm_to_set(fields, SC_queue_sort_method);
   }
   else {
      str = get_conf_value(&alp, *clpp, CF_name, CF_value, "sort_seq_no");
      lFreeList(&alp);
      if (!str) {
         SGE_ADD_MSG_ID(sprintf(SGE_EVENT, MSG_SCHEDCONF_INVALIDVALUEXFORQUEUESORTMETHOD_S, str));
         answer_list_add(alpp, SGE_EVENT, STATUS_ESYNTAX, ANSWER_QUALITY_ERROR);
         DEXIT;
         return -1; 
      }
      ul = !strcasecmp(str, "true") ? QSM_SEQNUM:QSM_LOAD;
      lSetUlong(ep, SC_queue_sort_method, ul);
      lDelElemStr(clpp, CF_name, "sort_seq_no");
      add_nm_to_set(fields, SC_queue_sort_method);
   }
      
   /* ---------  SC_job_load_adjustments */
   if (!set_conf_deflist(alpp, clpp, fields, "job_load_adjustments", ep, 
         SC_job_load_adjustments, CE_Type, intprt_as_load_adjustment)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_load_adjustment_decay_time */
   if (!set_conf_timestr(alpp, clpp, fields, "load_adjustment_decay_time", ep, 
                              SC_load_adjustment_decay_time)) {
      DEXIT;
      return -1;
   }


   /* --------- SC_load_formula */
   if (!set_conf_string(alpp, clpp, fields, "load_formula", ep, SC_load_formula)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_schedd_job_info */
   if (!set_conf_string(alpp, clpp, fields, "schedd_job_info", ep, SC_schedd_job_info)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_max_reservation */
   if (!set_conf_ulong(alpp, clpp, fields, "max_reservation", ep, SC_max_reservation)) {
      DEXIT;
      return -1;
   }

   /* --------- SC_max_reservation */
   if (!set_conf_string(alpp, clpp, fields, "default_duration", ep, SC_default_duration)) {
      DEXIT;
      return -1;
   }   
   
   DEXIT;
   return 0;
}

/****
 **** cull_read_in_schedd_conf
 ****/
lListElem *cull_read_in_schedd_conf(char *dirname, const char *filename,
                                    int spool, int *tag) 
{
   lListElem *ep;
   struct read_object_args args = { SC_Type, "schedd_conf", read_schedd_conf_work };
   int intern_tag = 0;

   DENTER(TOP_LAYER, "cull_read_in_schedd_conf");

   ep = read_object(dirname, filename, spool, 0, 0, &args, tag?tag:&intern_tag, NULL);

   DEXIT;
   return ep;
}


/***************************************************
 Read scheduler configuration

 spool: from spooled file (may contain additional fields)
 ***************************************************/
lList *read_sched_configuration(const char *common_dir, const char *fname, int spool, lList **alpp) 
{
   lList *confl = NULL;
   lListElem *ep;
   int write_default_config = 0;
   SGE_STRUCT_STAT st;

   DENTER(TOP_LAYER, "read_sched_configuration");

   if (!SGE_STAT(fname, &st)) {
      ep = cull_read_in_schedd_conf(NULL, fname, spool, NULL);
   } else {
      write_default_config = 1;
      ep = sconf_create_default();
   }

   if (ep) {
      confl = lCreateList("scheduler config", SC_Type);
      lAppendElem(confl, ep);
   } else {
      CRITICAL((SGE_EVENT, MSG_SCHEDCONF_CANTCREATESCHEDULERCONFIGURATION));
      answer_list_add(alpp, SGE_EVENT, STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR);
      DEXIT;
      return NULL;
   }

   if (write_default_config) {
      if (write_sched_configuration(1, 2, common_dir, lFirst(confl)) == NULL) {
         answer_list_add(alpp, MSG_SCHEDCONF_CANTCREATESCHEDULERCONFIGURATION, STATUS_ESEMANTIC, ANSWER_QUALITY_ERROR);
         DEXIT;
         return NULL;
      }
   }

   DEXIT;
   return confl;
}


